pkg/controller/migplan/migplan_controller.go

```diff
@@ -423,20 +423,80 @@ func (r *ReconcileMigPlan) planSuspended(plan *migapi.MigPlan) error {
 	return nil
 }
 
-func (r *ReconcileMigPlan) ensureMigrationStateLabels(plan *migapi.MigPlan) error {
+func (r *ReconcileMigPlan) ensureMigplanLabelsForMigMigrations(plan *migapi.MigPlan) error {
 
-	//Add or Update labels for migplan having running migrations
-	err := r.AddUpdateMigrationStateToPlan(plan, migctl.Running)
+	// ensure labels for migplan having running migrations
+	err := r.ensureMigPlanRunningLabel(plan)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
 
+	// ensure labels for migplan having failed migrations
+	err = r.ensureMigPlanFailedLabel(plan)
 	if err != nil {
 		return liberr.Wrap(err)
 	}
+	return nil
+}
 
-	//Add or Update labels for migplan having failed migrations
-	err = r.AddUpdateMigrationStateToPlan(plan, migctl.Failed)
+func (r ReconcileMigPlan) ensureMigPlanRunningLabel(plan *migapi.MigPlan) error {
+	migrationStateInstances := 0
+	migrations, err := plan.ListMigrations(r)
 	if err != nil {
 		return liberr.Wrap(err)
 	}
+
+	for _, m := range migrations {
+		if m.Status.HasCondition(migctl.Running) {
+			migrationStateInstances++
+			break
+		}
+	}
+
+	if migrationStateInstances > 0 {
+		// migplan has atleast 1 migmigration running
+		if plan.Labels == nil {
+			plan.Labels = make(map[string]string)
+		}
+		plan.Labels[migapi.MigplanMigrationRunning] = "true"
+		return nil
+	}
+	if plan.Labels != nil {
+		// migplan does not have any running migmigration remove the label
+		delete(plan.Labels, migapi.MigplanMigrationRunning)
+	}
+
+	return nil
+}
+
+func (r ReconcileMigPlan) ensureMigPlanFailedLabel(plan *migapi.MigPlan) error {
+	migrationStateInstances := 0
+	migrations, err := plan.ListMigrations(r)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	for _, m := range migrations {
+		if m.Status.HasCondition(migctl.Failed) || m.Status.HasCondition(migctl.PlanNotReady) {
+			migrationStateInstances++
+			break
+		}
+	}
+
+	if migrationStateInstances > 0 {
+		// migplan has atleast 1 migmigration running
+		if plan.Labels == nil {
+			plan.Labels = make(map[string]string)
+		}
+		plan.Labels[migapi.MigplanMigrationFailed] = "true"
+		return nil
+	}
+
+	if plan.Labels != nil {
+		// migplan does not have any failed migmigration remove the label
+		delete(plan.Labels, migapi.MigplanMigrationFailed)
+	}
+
```

Comment:
Same, missing update call for migplan ?

Reply:
Same https://github.com/alaypatel07/mig-controller/blob/efda3876f50953da4feb19c6ca6473d15b6e4929/pkg/controller/migplan/migplan_controller.go#L321